src/celengine/galaxy.cpp

```diff
@@ -8,114 +8,357 @@
     pixel[0] = 255;//65;
     pixel[1] = 255;//64;
     pixel[2] = 255;//65;
     pixel[3] = pixVal;
 }
 
-static void ColorTextureEval(float u, float /*v*/, float /*w*/, unsigned char *pixel)
+
+void colorTextureEval(float u, float /*v*/, float /*w*/, unsigned char *pixel)
 {
-    unsigned int i = (u*0.5f + 0.5f)*255.99f; // [-1, 1] -> [0, 255]
+    unsigned int i = static_cast<unsigned int>((static_cast<float>(u)*0.5f + 0.5f)*255.99f); // [-1, 1] -> [0, 255]
 
     // generic Hue profile as deduced from true-color imaging for spirals
     // Hue in degrees
-    float hue = 25 * tanh(0.0615f * (27 - i));
-    if (i >= 28) hue += 220;
+    float hue = 25.0f * std::tanh(0.0615f * (27.0f - static_cast<float>(i)));
+    if (i >= 28) hue += 220.0f;
     //convert Hue to RGB
     float r, g, b;
     DeepSkyObject::hsv2rgb(&r, &g, &b, hue, 0.20f, 1.0f);
-    pixel[0] = (unsigned char) (r * 255.99f);
-    pixel[1] = (unsigned char) (g * 255.99f);
-    pixel[2] = (unsigned char) (b * 255.99f);
+    pixel[0] = static_cast<unsigned char>(r * 255.99f);
+    pixel[1] = static_cast<unsigned char>(g * 255.99f);
+    pixel[2] = static_cast<unsigned char>(b * 255.99f);
+}
+
+
+struct GalaxyVertex
+{
+    Eigen::Vector4f position;
+    Eigen::Matrix<GLushort, 4, 1> texCoord; // texCoord.x = x, texCoord.y = y, texCoord.z = color index, texCoord.w = alpha
+};
+
+
+void draw(const GalaxyVertex *v, std::size_t count, const GLushort *indices)
+{
+    glVertexAttribPointer(CelestiaGLProgram::VertexCoordAttributeIndex,
+                          4, GL_FLOAT, GL_FALSE,
+                          sizeof(GalaxyVertex), v[0].position.data());
+    glVertexAttribPointer(CelestiaGLProgram::TextureCoord0AttributeIndex,
+                          4, GL_UNSIGNED_SHORT, GL_FALSE,
+                          sizeof(GalaxyVertex), v[0].texCoord.data());
+    glDrawElements(GL_TRIANGLES, count, GL_UNSIGNED_SHORT, indices);
+}
+
+
+GalaxyVertex *g_vertices = nullptr;
+GLushort *g_indices = nullptr;
+constexpr const std::size_t maxPoints = 8192; // 256k buffer
+
+
+std::optional<GalacticForm> buildGalacticForm(const fs::path& filename)
+{
+    Blob b;
+    BlobVector galacticPoints;
+
+    // Load templates in standard .png format
+    int width, height, rgb, j = 0, kmin = 9;
+    unsigned char value;
+    float h = 0.75f;
+    Image* img = LoadImageFromFile(filename);
+    if (img == nullptr)
+    {
+        celestia::util::GetLogger()->error("The galaxy template *** {} *** could not be loaded!\n", filename);
+        return std::nullopt;
+    }
+    width  = img->getWidth();
+    height = img->getHeight();
+    rgb    = img->getComponents();
+
+    auto& rng = celmath::getRNG();
+    for (int i = 0; i < width * height; i++)
+    {
+        value = img->getPixels()[rgb * i];
+        if (value > 10)
+        {
+            float x, y, z, r2, yy, prob;
+            z  = std::floor(static_cast<float>(i) / static_cast<float>(width));
```

Comment:
because of all casts it's hard to read such code, lets add temporary variables to hold cast values